golang/cmd/grafana-auth/http.go

```diff
@@ -0,0 +1,248 @@
+			panic(err)
+		}
+	}(closer)
+	defer reporter.Close()
+	opentracing.SetGlobalTracer(tracer)
+
+	// create the middleware
+	p := ginopentracing.OpenTracer([]byte("api-request-"))
+
+	// tell gin to use the middleware
+	router.Use(p)
+
+	// Healthcheck
+	router.GET("/", func(c *gin.Context) {
+		c.String(http.StatusOK, "online")
+	})
+
+	// Version of the API
+	v1 := router.Group("/api/v1")
+	{
+		v1.GET("/:service/*data", getProxyHandler)
+	}
+
+	err = router.Run(":80")
+	if err != nil {
+		panic(err)
+	}
+}
+
+func handleInvalidInputError(parentSpan opentracing.Span, c *gin.Context, err error) {
+
+	ext.LogError(parentSpan, err)
+	traceID, _ := internal.ExtractTraceID(parentSpan)
+
+	zap.S().Errorw("Invalid input error",
+		"error", err,
+		"trace id", traceID,
+	)
+
+	c.String(400, "You have provided a wrong input. Please check your parameters and mention the following trace id while contacting our support: "+traceID)
+}
+
+type getProxyRequest struct {
+	Service     string `uri:"service" binding:"required"`
+	OriginalURI string `uri:"data" binding:"required"`
+}
+
+func getProxyHandler(c *gin.Context) {
+	// Jaeger tracing
+	var span opentracing.Span
+	if cspan, ok := c.Get("tracing-context"); ok {
+		span = ginopentracing.StartSpanWithParent(cspan.(opentracing.Span).Context(), "getProxyHandler", c.Request.Method, c.Request.URL.Path)
+	} else {
+		span = ginopentracing.StartSpanWithHeader(&c.Request.Header, "getProxyHandler", c.Request.Method, c.Request.URL.Path)
+	}
+	defer span.Finish()
+
+	var getProxyRequest getProxyRequest
+	var err error
+
+	// Failed to parse request into service name and original url
+	err = c.BindUri(&getProxyRequest)
+	if err != nil {
+		handleInvalidInputError(span, c, err)
+		return
+	}
+
+	// Grafana session not present in request
+	session, err := c.Cookie("grafana_session")
+	if err != nil {
+
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	// Check if user is logged in
+	loggedIn, err := CheckUserLoggedIn(session)
+	if err != nil {
+		handleInvalidInputError(span, c, err)
+	}
+
+	// Abort if not logged in
+	if !loggedIn {
+		c.AbortWithStatus(http.StatusForbidden)
+		return
+	}
+
+	// Switch to handle our services
+	switch getProxyRequest.Service {
+	case "factoryinput":
+		HandleFactoryInput(c, session, getProxyRequest)
+	default:
+		c.AbortWithStatus(http.StatusBadRequest)
+	}
+}
+
+// HandleFactoryInput handles proxy requests to factoryinput
+func HandleFactoryInput(c *gin.Context, sessioncookie string, request getProxyRequest) {
+	// Jaeger tracing
+	var span opentracing.Span
+	if cspan, ok := c.Get("tracing-context"); ok {
+		span = ginopentracing.StartSpanWithParent(cspan.(opentracing.Span).Context(), "HandleFactoryInput", c.Request.Method, c.Request.URL.Path)
+	} else {
+		span = ginopentracing.StartSpanWithHeader(&c.Request.Header, "HandleFactoryInput", c.Request.Method, c.Request.URL.Path)
+	}
+	defer span.Finish()
+
+	proxyUrl := strings.TrimPrefix(string(request.OriginalURI), "/")
```

Comment:
Use Service name in k8s instead of user provided input for the URL

Reply:
Fixed in 119c22ffd55f371187c14fa3ebd11085eb7235f3